Public/Src/Engine/Processes/SharedOpaqueOutputLogger.cs

```diff
@@ -0,0 +1,238 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.ContractsLight;
+using System.Globalization;
+using System.IO;
+using System.Linq;
+using BuildXL.Pips.Operations;
+using BuildXL.Utilities;
+using BuildXL.Utilities.Collections;
+using JetBrains.Annotations;
+
+namespace BuildXL.Processes
+{
+    /// <summary>
+    /// Responsible for keeping a log of all paths written to by a given process pip.
+    /// 
+    /// Paths are flushed to an underlying file as soon as they're reported via 
+    /// <see cref="RecordFileWrite(PathTable, AbsolutePath)"/>.
+    /// 
+    /// A number of root directories may be set in which case the log will record only 
+    /// those paths that fall under one of those directories.  The typical use case is to
+    /// use all shared opaque directory outputs of a process as root directories.
+    /// </summary>
+    /// <remarks>
+    /// NOTE: not thread-safe
+    /// 
+    /// NOTE: must be serializable in order to be compatible with VM execution; for this 
+    ///       reason, this class must not have a field of type <see cref="PathTable"/>.
+    /// </remarks>
+    public sealed class SharedOpaqueOutputLogger : IDisposable

```

Comment:
>SharedOpaqueOutputLogger [](start = 24, length = 24)

is there any logic in place to prevent us from file corruption? Thinking in the case where the writer dies half way through writing to the sideband file. And a next build trying to read from it and getting some unexpected/malformed paths

Reply:
There isn't at the moment.  I'll look into using FileEnvelope, as Iman suggested.

Reply:
Added FileEnvelope for checking file integrity.  However, there is an option to read whatever we can even if the file wasn't closed cleanly, because that was one of the indented use cases for this to begin with (e.g., bxl crashes, doesn't mark shared opaque, and doesn't cleanly close sideband files either)